text/0000-edition-based-method-disambiguation.md

```diff
@@ -0,0 +1,224 @@
+* A new inherent method on an pre-edition type (no recent examples)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Disambiguation can require invasive changes
+
+In simple cases where you only call a single method in an expression switching
+from the ambiguous method call syntax, `self.method(args...)` to the
+unambiguous function call syntax `Trait::method(self, args...)` is an easy
+change. In longer method call chains however there isn't a way to disambiguate
+the trait a method is associated with when calling it without splitting up that
+expression into multiple expressions, which can change drop behavior and
+prevent temporaries for living as long as they need to.
+
+This RFC intentionally avoids solving this problem or even proposing strawmen
+versions of the syntax to avoid distracting from the core issue, but at the
+same time it increases the need for a language syntax extension for quickly
+disambiguating the trait a method call should come from.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+[This
+comment](https://github.com/rust-lang/rust/issues/88967#issuecomment-938024847)
+on the `Iterator::intersperse` issue details a few alternatives that were
+discussed by the libs team when we encountered that specific issue. These
+include both short term and long term suggestions.
+
+**Short Term**
+
+* use build scripts in crates where this breakage is expected to detect when
+  the breakage is introduced by the compiler and disable the conflicting APIs
+  internally using `#[cfg]` directives.
+* use `#[cfg(accessible)]` in the crate where the breakage is expected to
+  automatically remove the conflicting API when the upstream API is introduced.
+
+These solutions don't solve the problem generally, and instead address the
+specific breakage that are known or expected. We can and do catch many such
+issues via crater but we cannot test all crates via crater and we still end up
+breaking people's crates on nightly before the crater runs have had a chance to
+catch any breakage.
+
+**Longer Term**
+
+* encoding the released std version in crates published to crates.io when
+  uploading and ignore methods introduced in newer releases when building that
+  published dependency.
+* [Supertrait item shadowing](https://github.com/rust-lang/rfcs/pull/2845)
+
+These solutions also do not solve the problem generally. The former solution
+would only solve the problem for crates uploaded to crates.io, and the latter
+would only solve the problem for traits where the ambiguity is introduced by a
+supertrait.
+
+**Note**: While the "Supertrait item shadowing" rfc is listed here as an
+alternative it should not necessarily be seen as mutually exclusive, and could
+absolutely be used as a complement to this feature. Many of these warnings
+could be ambiguity warnings could be resolved by trait precidence and edition
+based fallback would only be encountered for the rarer cases of unrelated name
+collisions.
```

Comment:
Does this mean that `#[cfg(accessible)]` and `#[cfg(version)]` _are_ mutually exclusive with this feature? The interactions between this feature and the `cfg` annotations are not yet clear to me. What should happen when both are used? I would find it helpful if the interactions between the two was spelled out, with guidance on when each approach should be preferred.

Reply:
> Does this mean that `#[cfg(accessible)]` and `#[cfg(version)]` _are_ mutually exclusive with this feature?

Ah, no, I don't think these would be mutually exclusive features either, I'll clarify this note.